server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -335,15 +378,22 @@ public void onResponse(ShardSnapshotResult shardSnapshotResult) {
             @Override
             public void onFailure(Exception e) {
                 final String failure;
+                final Stage nextStage;
                 if (e instanceof AbortedSnapshotException) {
+                    nextStage = Stage.FAILURE;
                     failure = "aborted";
                     logger.debug(() -> format("[%s][%s] aborted shard snapshot", shardId, snapshot), e);
+                } else if (e instanceof PausedSnapshotException) {
+                    nextStage = Stage.PAUSED;
+                    failure = "paused for removal of node holding primary";
+                    logger.debug(() -> format("[%s][%s] pausing shard snapshot: [%s]", shardId, snapshot), e);
                 } else {
+                    nextStage = Stage.FAILURE;
                     failure = summarizeFailure(e);
                     logger.warn(() -> format("[%s][%s] failed to snapshot shard", shardId, snapshot), e);
                 }
-                snapshotStatus.moveToFailed(threadPool.absoluteTimeInMillis(), failure);
-                notifyFailedSnapshotShard(snapshot, shardId, failure, snapshotStatus.generation());
+                final var shardState = snapshotStatus.moveToUnsuccessful(nextStage, failure, threadPool.absoluteTimeInMillis());
+                notifyFailedSnapshotShard(snapshot, shardId, shardState, failure, snapshotStatus.generation());
```

Comment:
Nit: Can we rename this method to `notifyUnsuccessfulSnapshotShard` to match `moveToUnsuccessful`?